### PR TITLE
smc: fix discrete sampling

### PR DESCRIPTION
**Pull request by aloctavodia**
Following [discourse](https://discourse.pymc.io/t/bug-nan-smc-weights-and-indexoutofboundserror-in-theano/2092/4), this fix a problem with discrete variables.

### CODE REVIEW

**Comment by twiecki** [comment]
Can we add a test so this doesn't reappear?

**Comment by aloctavodia** [comment]
@twiecki  May I mark this with 

```python
@pytest.mark.xfail(condition=(theano.config.floatX == "float32"), reason="Fails on float32")
```

File: pymc3/tests/test_smc.py [hunk 1]
```diff
@@ -47,6 +49,15 @@ def test_sample(self):
         mu1d = np.abs(x).mean(axis=0)
         np.testing.assert_allclose(self.muref, mu1d, rtol=0., atol=0.03)
 
+    @pytest.mark.xfail(condition=(theano.config.floatX == "float32"), reason="Fails on float32")
```
Comment (twiecki) [inline pymc3/tests/test_smc.py]: How could this fail on float32, it's just running a model?

File: pymc3/step_methods/smc.py [hunk 2]
```diff
@@ -177,16 +188,18 @@ def _initial_population(draws, model, variables):
...
     start = model.test_point
+    dict_to_array = DictToArrayBijection(ArrayOrdering(variables), start).map
     init_rnd = pm.sample_prior_predictive(draws, model=model)
     for v in variables:
         var_info[v.name] = (start[v.name].shape, start[v.name].size)
 
     for i in range(draws):
         point = pm.Point({v.name: init_rnd[v.name][i] for v in variables}, model=model)
-        population.append(model.dict_to_array(point))
+        population.append(dict_to_array(point))
```
Comment (twiecki) [inline pymc3/step_methods/smc.py]: Where is the actual fix here?

Comment (aloctavodia) [reply pymc3/step_methods/smc.py]: `model.dict_to_array` returns an empty array for discrete variables (bijection method is reading only continuous variables). Thus I basically create a version that reads all variables.

File: pymc3/tests/test_smc.py (hunk 1 above)
Comment (aloctavodia) [reply pymc3/tests/test_smc.py]: Therei is a TypeError: expected type_num 11 (NPY_FLOAT32) got 12 error when evaluating the likelihood_logp function https://travis-ci.org/pymc-devs/pymc3/jobs/447528838#L1870

Comment (twiecki) [reply pymc3/tests/test_smc.py]: You probably need to floatX the observed data.

File: pymc3/step_methods/smc.py (hunk 2 above)
Comment (twiecki) [reply pymc3/step_methods/smc.py]: Isn't that bug then upstream of this? Why should `model.dict_to_array` only return continuous variables in the first place?

File: pymc3/tests/test_smc.py (hunk 1 above)
Comment (aloctavodia) [reply pymc3/tests/test_smc.py]: Right, thanks! This will fix it.

File: pymc3/step_methods/smc.py (hunk 2 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc.py]: not sure, I asked that in slack.

Comment (aloctavodia) [reply pymc3/step_methods/smc.py]: @twiecki what do you think we should do?

**Comment by twiecki** [comment]
Thanks @aloctavodia!

**Pull request merged**